File: dart/lib/lunar_api_client.dart
import 'dart:convert';
import 'dart:io';

class LunarApiResponse {
  final int status;
  final Map<String, dynamic> json;
  final int? dailyLimit;
  final int? dailyRemaining;
  final int? dailyResetEpochSeconds;
  const LunarApiResponse(this.status, this.json, this.dailyLimit, this.dailyRemaining, this.dailyResetEpochSeconds);
}

class LunarApiException implements Exception {
  final int status;
  final Map<String, dynamic> json;
  const LunarApiException(this.status, this.json);
  @override String toString() => 'Vietnamese Lunar API returned HTTP $status';
}

/// Transport-only Dart/Flutter client. It contains no calendar algorithm.
class LunarApiClient {
  final String apiKey;
  final String baseUrl;
  final Duration timeout;
  final HttpClient _http;

  LunarApiClient(this.apiKey, {this.baseUrl = 'https://lunar.ivadev.workers.dev', this.timeout = const Duration(seconds: 10), HttpClient? httpClient})
      : assert(apiKey != ''), _http = httpClient ?? HttpClient();

  Future<LunarApiResponse> toLunar(String date, {String? profile}) => _get('/v1/lunar', {'date': date, if (profile != null) 'profile': profile});
  Future<LunarApiResponse> toSolar(int day, int month, int year, {bool isLeapMonth = false, String? profile}) => _get('/v1/solar', {'day': '$day', 'month': '$month', 'year': '$year', 'leap': '$isLeapMonth', if (profile != null) 'profile': profile});
  Future<LunarApiResponse> solarTerms(int year, {String? profile}) => _get('/v1/solar-terms', {'year': '$year', if (profile != null) 'profile': profile});

  Future<LunarApiResponse> toLunarBatch(List<String> dates, {String? profile}) async {
    if (dates.isEmpty || dates.length > 31) throw ArgumentError('dates must contain 1 to 31 values');
    final uri = Uri.parse(baseUrl).resolve('/v1/lunar/batch');
    final request = await _http.postUrl(uri).timeout(timeout);
    request.headers.set(HttpHeaders.acceptHeader, 'application/json');
    request.headers.set(HttpHeaders.contentTypeHeader, 'application/json');
    request.headers.set('X-API-Key', apiKey);
    request.headers.set('X-Client-Version', 'dart/1.0.0');
    request.write(jsonEncode({'dates': dates, if (profile != null) 'profile': profile}));
    return _read(await request.close().timeout(timeout));
  }

  Future<LunarApiResponse> _get(String path, Map<String, String> query) async {
    final uri = Uri.parse(baseUrl).resolve(path).replace(queryParameters: query);
    final request = await _http.getUrl(uri).timeout(timeout);
    request.headers.set(HttpHeaders.acceptHeader, 'application/json');
    request.headers.set('X-API-Key', apiKey);
    request.headers.set('X-Client-Version', 'dart/1.0.0');
    return _read(await request.close().timeout(timeout));
  }

  Future<LunarApiResponse> _read(HttpClientResponse response) async {
    final text = await utf8.decoder.bind(response).join();
    final body = text.isEmpty ? <String, dynamic>{} : jsonDecode(text) as Map<String, dynamic>;
    if (response.statusCode < 200 || response.statusCode >= 300) throw LunarApiException(response.statusCode, body);
    int? header(String name) => int.tryParse(response.headers.value(name) ?? '');
    return LunarApiResponse(response.statusCode, body, header('X-RateLimit-Daily-Limit'), header('X-RateLimit-Daily-Remaining'), header('X-RateLimit-Daily-Reset'));
  }

  void close() => _http.close();
}

File: kotlin/LunarApiClient.kt
package io.ivadev.lunar.api

import java.io.IOException
import java.net.HttpURLConnection
import java.net.URI
import java.net.URLEncoder
import java.nio.charset.StandardCharsets

data class LunarApiResponse(
    val status: Int,
    val json: String,
    val dailyLimit: Int?,
    val dailyRemaining: Int?,
    val dailyResetEpochSeconds: Long?
)

class LunarApiException(val status: Int, val responseJson: String) : IOException("Vietnamese Lunar API returned HTTP $status")

/** Transport-only JVM/Android client. It contains no calendar algorithm. */
class LunarApiClient(
    private val apiKey: String,
    private val baseUrl: String = "https://lunar.ivadev.workers.dev",
    private val connectTimeoutMillis: Int = 10_000,
    private val readTimeoutMillis: Int = 10_000
) {
    init { require(apiKey.isNotBlank()) { "apiKey is required" } }

    fun toLunar(date: String, profile: String? = null) = get("/v1/lunar?date=${encode(date)}${profilePart(profile)}")

    fun toSolar(day: Int, month: Int, year: Int, isLeapMonth: Boolean = false, profile: String? = null) =
        get("/v1/solar?day=$day&month=$month&year=$year&leap=$isLeapMonth${profilePart(profile)}")

    fun solarTerms(year: Int, profile: String? = null) = get("/v1/solar-terms?year=$year${profilePart(profile)}")

    fun toLunarBatch(dates: List<String>, profile: String? = null): LunarApiResponse {
        require(dates.size in 1..31) { "dates must contain 1 to 31 values" }
        val escaped = dates.joinToString(",") { "\"${jsonEscape(it)}\"" }
        val body = "{\"dates\":[$escaped]${profile?.let { ",\"profile\":\"${jsonEscape(it)}\"" } ?: ""}}"
        return request("/v1/lunar/batch", "POST", body)
    }

    private fun get(path: String) = request(path, "GET", null)

    private fun request(path: String, method: String, body: String?): LunarApiResponse {
        val connection = URI.create(baseUrl.trimEnd('/') + path).toURL().openConnection() as HttpURLConnection
        connection.requestMethod = method
        connection.connectTimeout = connectTimeoutMillis
        connection.readTimeout = readTimeoutMillis
        connection.setRequestProperty("Accept", "application/json")
        connection.setRequestProperty("X-API-Key", apiKey)
        connection.setRequestProperty("X-Client-Version", "kotlin/1.0.0")
        if (body != null) {
            connection.doOutput = true
            connection.setRequestProperty("Content-Type", "application/json")
            connection.outputStream.use { it.write(body.toByteArray(StandardCharsets.UTF_8)) }
        }
        val status = connection.responseCode
        val stream = if (status in 200..299) connection.inputStream else connection.errorStream
        val json = stream?.bufferedReader(StandardCharsets.UTF_8)?.use { it.readText() } ?: ""
        val result = LunarApiResponse(status, json, connection.intHeader("X-RateLimit-Daily-Limit"), connection.intHeader("X-RateLimit-Daily-Remaining"), connection.longHeader("X-RateLimit-Daily-Reset"))
        connection.disconnect()
        if (status !in 200..299) throw LunarApiException(status, json)
        return result
    }

    private fun profilePart(profile: String?) = profile?.let { "&profile=${encode(it)}" } ?: ""
    private fun encode(value: String) = URLEncoder.encode(value, StandardCharsets.UTF_8.name())
    private fun jsonEscape(value: String) = value.replace("\\", "\\\\").replace("\"", "\\\"")
    private fun HttpURLConnection.intHeader(name: String) = getHeaderField(name)?.toIntOrNull()
    private fun HttpURLConnection.longHeader(name: String) = getHeaderField(name)?.toLongOrNull()
}
